File: 2_term/2/2-2/arrayStack.cpp
#include "arrayStack.h"

ArrayStack::ArrayStack() : capacity(2), size (0), array(new double[2])
{
}

ArrayStack::~ArrayStack()
{
    delete[] array;
}

void ArrayStack::push(double number)
{
    if (size == capacity)
    {
        double* newArray = new double[capacity * 2];
        for (int i = 0; i < capacity; i++)
        {
            newArray[i] = array[i];
        }
        capacity = capacity * 2;
        delete[] array;
        array = newArray;
    }
    for (int i = size; i > 0; i--)
    {
        array[i] = array[i - 1];
    }
    array[0] = number;
    size = size + 1;
}

double ArrayStack::pop()
{
    int result = array[0];
    for (int i = 0; i < size; i++)
    {
        array[i] = array[i + 1];
    }
    size = size - 1;
    array[size] = 0;
    if (size == capacity/ 2 &&
            capacity % 2 == 0 && capacity / 2 != 0)
    {
        double* newArray = new double[size];
        for (int i = 0; i < size; i++)
        {
            newArray[i] = array[i];
        }
        capacity = capacity / 2;
        delete[] array;
        array = newArray;
    }
    return result;
}

double ArrayStack::top()
{
    return array[0];
}

bool ArrayStack::isEmpty() const
{
    return size == 0;
}

File: 2_term/2_C#/2-4/2-4/ArrayStack.cs
using System;

namespace Application
{
    public class ArrayStack<T>: IStack<T>
    {
        private T[] array = new T[2];
        private int capacity = 2;
        private int size = 0;

        public void Push(T number)
        {
            if (size == capacity)
            {
                Array.Resize(ref array, capacity * 2);
                capacity = capacity * 2;
            }
            for (int i = capacity - 1; i > 0; i--)
            {
                array[i] = array[i - 1];
            }
            array[0] = number;
            size = size + 1;
        }

        public T Pop()
        {
            T result = array[0];
            size = size - 1;
            for (int i = 0; i < capacity - 1; i++)
            {
                array[i] = array[i + 1];
            }
            if (size == capacity / 2 &&
                capacity % 2 == 0 && capacity / 2 != 0)
            {
                Array.Resize(ref array, capacity / 2);
                capacity = capacity / 2;
            }
            return result;
        }

        public T Top()
        {
            return array[0];
        }

        public bool IsEmpty()
        {
            return size == 0;
        }
    }
}

